server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java

```diff
@@ -73,19 +80,49 @@ public void handle(Request request, Response response) throws Exception {
     JsonWriter jsonWriter = response.newJsonWriter();
 
     jsonWriter.beginObject();
-    writeInstalling(jsonWriter, compatiblePluginsByKey);
-    writeRemoving(jsonWriter, compatiblePluginsByKey);
+    writePlugins(jsonWriter, compatiblePluginsByKey);
     jsonWriter.endObject();
     jsonWriter.close();
   }
 
-  private void writeInstalling(JsonWriter json, Map<String, Plugin> compatiblePluginsByKey) {
-    Collection<PluginInfo> plugins = pluginDownloader.getDownloadedPlugins();
-    pluginWSCommons.writePluginInfoList(json, plugins, compatiblePluginsByKey, ARRAY_INSTALLING);
+  private void writePlugins(JsonWriter json, Map<String, Plugin> compatiblePluginsByKey) {
+    Collection<PluginInfo> uninstalledPlugins = installer.getUninstalledPlugins();
+    pluginWSCommons.writePluginInfoList(json, uninstalledPlugins, compatiblePluginsByKey, ARRAY_REMOVING);
+
+    Collection<PluginInfo> downloadedPlugins = pluginDownloader.getDownloadedPlugins();
+    Collection<PluginInfo> installedPlugins = installer.getPluginInfos();
+    List<String> installedPluginKeys = from(installedPlugins).transform(PluginInfoToKey.INSTANCE).toList();
+
+    Collection<PluginInfo> newPlugins = from(downloadedPlugins)
```

Comment:
I think I prefer one loop that feeds 2 lists `newPlugins` and `updatedPlugins`. You could call the `MatchPluginKeys` predicate explicitly of just create a test method.